### PR TITLE
Add Procedure call support for sql connector

### PR DESCRIPTION
**Pull request by niveathika**
## Purpose
$subject

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/24412, https://github.com/ballerina-platform/ballerina-lang/issues/12804

## Approach
Design Doc
https://docs.google.com/document/d/1d0-yLkJdvI4rZo4mKd5ckizWCGT3wPRZqRHEWpxz75I/edit#heading=h.x7zyie9d96p0

## Samples
```ballerina
int uid = <@untainted> id;
sql:OutParameter insertId = new;

sql:ProcedureCallResult result = check mysqlClient->call(`call InsertPerson(${uid}, ${insertId};`);
io:println(insertId.get(int));
```

## Remarks
Needs to be verified against
- [x] H2
- [x] MySQL
- [x] MSSQL
- [x] Postgre
- [x] Oracle 11
- [x] Oracle 12
Added testscases for MySQL, manually tested for others.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [x] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [x] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [x] Added necessary documentation  
  - [x] API documentation 
  - [x] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: stdlib/database/jdbc/src/main/ballerina/src/java.jdbc/client.bal [hunk 1]
```diff
@@ -96,6 +96,15 @@ public type Client client object {
         }
     }
 
+    public remote function call(@untainted sql:ParameterizedCallQuery sqlQuery, typedesc<record {}>[] rowTypes = [])
```
Comment (daneshk) [inline stdlib/database/jdbc/src/main/ballerina/src/java.jdbc/client.bal:99]: Add API docs

Comment (niveathika) [reply stdlib/database/jdbc/src/main/ballerina/src/java.jdbc/client.bal:105]: Added in 0c13c50

File: stdlib/database/jdbc/src/main/ballerina/src/java.jdbc/client.bal [hunk 2]
```diff
@@ -96,6 +96,21 @@ public type Client client object {
         }
     }
 
+    # Executes a SQL stored procedure and returns the result as stream and execution summary.
+    #
+    # + sqlQuery - The query to executed the SQL stored procedure.
```
Comment (anupama-pathirage) [inline stdlib/database/jdbc/src/main/ballerina/src/java.jdbc/client.bal:101]:
```suggestion
    # + sqlQuery - The query to executed by the SQL stored procedure.
```

File: stdlib/database/jdbc/src/main/ballerina/src/java.jdbc/client.bal [hunk 3]
```diff
@@ -96,6 +96,21 @@ public type Client client object {
...
+    # + rowTypes - The array of `typedesc` of the records that should be returned as a result. If this is not provided
+    #               the default column names of the query result set be used for the record attributes.
+    # + return - Summary of the execution is returned in `ProcedureCallResult` or `sql:Error`.
+    public remote function call(@untainted sql:ParameterizedCallQuery sqlQuery, typedesc<record {}>[] rowTypes = [])
```
Comment (anupama-pathirage) [inline stdlib/database/jdbc/src/main/ballerina/src/java.jdbc/client.bal:105]: The convention of using full stop is not properly followed in the api doc statements. Refer below sample by @MaryamZi . No full stop at the end of first sentence of the parameter and return type description and it is there only at function description. Maybe u can fix the newly added docs in this PR and sent another one to fix all the places in db modules in a separate PR as there is a lot of places to fix.  

```ballerina
# Description for the function.
#
# + i - One sentence only
# + s - Sentence one. Sentence two.
# + return - Return description
public function foo(int i, string s) returns boolean {
    return true;
}
```

File: stdlib/database/sql/src/main/java/org/ballerinalang/sql/utils/OutParameterUtils.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+                    }
+                    return convert((boolean) value, sqlType, ballerinaType, false);
+                case Types.REF:
+                case Types.STRUCT:
+                    return convert((Struct) value, sqlType, ballerinaType);
+                case Types.SQLXML:
+                    return convert((SQLXML) value, sqlType, ballerinaType);
+                default:
+                    // Cannot reach to Default
+                    return ErrorGenerator.getSQLApplicationError("Unsupported sql type " + sqlType);
```
Comment (anupama-pathirage) [inline stdlib/database/sql/src/main/java/org/ballerinalang/sql/utils/OutParameterUtils.java:138]:
```suggestion
                    return ErrorGenerator.getSQLApplicationError("Unsupported SQL type " + sqlType);
```

**Comment by anupama-pathirage** [comment]
@niveathika Can we add unit tests to test this functionality?

File: stdlib/database/sql/src/main/ballerina/src/sql/Module.md [hunk 5]
```diff
@@ -340,5 +340,30 @@ if (ret is error) {
...
+```ballerina
+int uid = 10;
+sql:OutParameter insertId = new;
+
+var ret = dbClient->call(`call InsertPerson(${uid}, ${insertId}`);
+if (ret is error) {
+    io:println("Error occurred:", err.message());
+} else {
+    io:println("Outparameter insert id: ", insertId.get(int));
+    stream<record{}, sql:Error>? resultStr = result.queryResult;
```
Comment (anupama-pathirage) [inline stdlib/database/sql/src/main/ballerina/src/sql/Module.md:357]: Is this working sample? What is `result` variable here?

File: stdlib/database/jdbc/src/main/ballerina/src/java.jdbc/client.bal (hunk 3 above)
Comment (niveathika) [reply stdlib/database/jdbc/src/main/ballerina/src/java.jdbc/client.bal]: Fixed in c8978fd

File: stdlib/database/sql/src/main/ballerina/src/sql/Module.md (hunk 5 above)
Comment (niveathika) [reply stdlib/database/sql/src/main/ballerina/src/sql/Module.md]: Seems to be an error in the sample, it should be `ret` variable returned from the call api. Fixed in c8978fd

File: stdlib/database/sql/src/main/java/org/ballerinalang/sql/utils/OutParameterUtils.java (hunk 4 above)
Comment (niveathika) [reply stdlib/database/sql/src/main/java/org/ballerinalang/sql/utils/OutParameterUtils.java]: Fixed in f56de73

File: stdlib/database/jdbc/src/main/ballerina/src/java.jdbc/client.bal (hunk 2 above)
Comment (niveathika) [reply stdlib/database/jdbc/src/main/ballerina/src/java.jdbc/client.bal]: Fixed in b21f0db

**Comment by niveathika** [comment]
> @niveathika Can we add unit tests to test this functionality?
Added in bc5fe54

File: stdlib/database/sql/src/main/ballerina/src/sql/Module.md [hunk 6]
```diff
@@ -340,5 +340,30 @@ if (ret is error) {
...
+#### Execute SQL stored procedures
+
+This example demonstrates how to execute a stored procedure with a single INSERT statement that is executed via the 
+`call` remote function of the client.
+
+```ballerina
+int uid = 10;
+sql:OutParameter insertId = new;
+
+var ret = dbClient->call(`call InsertPerson(${uid}, ${insertId}`);
```
Comment (daneshk) [inline stdlib/database/sql/src/main/ballerina/src/sql/Module.md:352]:
```suggestion
var ret = dbClient->call(`call InsertPerson(${uid}, ${insertId})`);
```

**Pull request merged**